db.go

```diff
@@ -227,9 +232,25 @@ func Open(dir string, l log.Logger, r prometheus.Registerer, opts *Options) (db
 		db.lockf = &lockf
 	}
 
-	db.compactor = newCompactor(dir, r, l, &compactorOptions{
-		maxBlockRange: opts.MaxBlockDuration,
-	})
+	copts := &compactorOptions{
+		blockRanges: defaultBlockRanges,
+	}
+	if copts.blockRanges[0] != int64(opts.MinBlockDuration) {
+		copts.blockRanges = append([]int64{int64(opts.MinBlockDuration)}, copts.blockRanges...)
+	}
+
+	for {
+		if len(copts.blockRanges) == 1 {
+			break
+		}
+
+		// Max overflow is restricted to 20%.
+		if float64(copts.blockRanges[len(copts.blockRanges)-1])/float64(opts.RetentionDuration) > 0.2 {
+			copts.blockRanges = copts.blockRanges[:len(copts.blockRanges)-1]
+		}
+	}
```

Comment:
This would just spin if the highest range is less than 20% of the retention, no?